### DISCUSSION TITLE
Getting error Responder trait not implemented

### DISCUSSION THREAD

**Discussion by sagarnayak** [rwf2/Rocket#1750] [Questions] [answered]
My Code -

	#[macro_use]
	extern crate rocket;
	extern crate rocket_contrib;

	use rand::Rng;
	use rocket_contrib::json::Json;
	use serde::{Deserialize, Serialize};

	#[derive(Serialize, Deserialize)]
	pub struct User {
		pub name: String,
	}

	#[derive(Serialize, Deserialize)]
	pub struct StatusMessage {
		pub code: i32,
		pub message: String,
	}

	#[get("/getJsonResponse")]
	fn get_json_response() -> Result<Json<User>, Json<StatusMessage>> {
		let random_result_to_send = rand::thread_rng().gen_bool(1.5);

		if random_result_to_send {
			Ok(
				Json(
					User {
						name: "User name".to_string()
					}
				)
			)
		} else {
			Err(
				Json(
					StatusMessage {
						code: 400,
						message: "got an error".to_string(),
					}
				)
			)
		}
	}

	#[launch]
	fn rocket() -> _ {
		let routes = routes![
			get_json_response
		];
		rocket::build()
			.mount("/", routes)
	}

roml -

	[package]
	name = "web_service"
	version = "0.1.0"
	edition = "2018"

	# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

	[dependencies]
	rocket = "0.5.0-rc.1"
	rand = "0.8.4"
	serde = { version = "1.0.64", features = ["derive"] }
	serde_json = "1.0.64"
	[dependencies.rocket_contrib]
	version = "0.4.10"
	default-features = false
	features = ["json"]

Error -

	error[E0277]: the trait bound `rocket_contrib::json::Json<User>: Responder<'_, '_>` is not satisfied
	  --> src\main.rs:21:27
	   |
	21 | fn get_json_response() -> Result<Json<User>, Json<StatusMessage>> {
	   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Responder<'_, '_>` is not implemented for `rocket_contrib::json::Json<User>`
	   |
	   = note: required because of the requirements on the impl of `Responder<'_, '_>` for `Result<rocket_contrib::json::Json<User>, rocket_contrib::json::Json<StatusMessage>>`
	   = note: required by `route::handler::<impl Outcome<rocket::Response<'o>, Status, rocket::Data<'o>>>::from`

	error[E0277]: the trait bound `rocket_contrib::json::Json<StatusMessage>: Responder<'_, '_>` is not satisfied
	  --> src\main.rs:21:27
	   |
	21 | fn get_json_response() -> Result<Json<User>, Json<StatusMessage>> {
	   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Responder<'_, '_>` is not implemented for `rocket_contrib::json::Json<StatusMessage>`
	   |
	   = note: required because of the requirements on the impl of `Responder<'_, '_>` for `Result<rocket_contrib::json::Json<User>, rocket_contrib::json::Json<StatusMessage>>`
	   = note: required by `route::handler::<impl Outcome<rocket::Response<'o>, Status, rocket::Data<'o>>>::from`

	error: aborting due to 2 previous errors

	For more information about this error, try `rustc --explain E0277`.
	error: could not compile `web_service`

Do I have to implement the responder my self or I am missing something?

**Comment by SergioBenitez** [comment, marked as answer]
There is no contrib in 0.5. Please read the CHANGELOG.

**Comment by sagarnayak** [comment]
This is what i have changed.

	[dependencies]
	rocket = { version = "0.5.0-rc.1", features = ["json"] }
	rand = "0.8.4"
	serde = { version = "1.0.64", features = ["derive"] }
	serde_json = "1.0.64"

Imports -

	#[macro_use]
	extern crate rocket;

	use rand::Rng;
	use serde::{Deserialize, Serialize};
	use rocket::serde::json::Json;

Thanks @SergioBenitez  its fixed.

**Reply by SergioBenitez** [reply]
You don't need to depend on `serde_json` directly. You also don't need to depend on `serde` at all if you add the following to structs where you derive Serialize or deserialize:

```
#[serde(crate = "rocket::serde")]
```

**Reply by sagarnayak** [reply]
Do you mean I can write like this -

	#[serde(crate = "rocket::serde")]
	struct User {
		name: String,
	}

and i can remote the serde_json = "1.0.64"

**Reply by SergioBenitez** [reply]
These are independent things. You can remove `serde_json` full-stop without doing anything else. To remove `serde`, however, you need to add that attribute wherever you derive Serialize or Deserialize.
